### ISSUE TITLE
[feature] Upload all binary packages from a lockfile

### ISSUE THREAD

**Issue by Inujel** [conan-io/conan#12885]
I often find myself in a situation where I want to upload all recipes and binary packages (to my on-premises conan server) that were built or used during a previous `conan create/install ... -b missing` command.

Either during development when I'm satisfied with a dependency graph and I want to save it. Or in CI when I want to be sure to save all dependencies that were used for a given build.

However, in both cases, I don't want to use `conan upload *`:
My development workstation's local cache may just contain anything and during CI, I usually cache `~/.conan/data`, so it might also contain much more packages than what I want to upload.

Does anyone else have the same problem? 

I think an elegant solution would be to upload all recipes and binary packages that are listed in a lockfile.
```shell
conan install . ... -b missing -if build/
conan upload -r my_remote --lockfile build/conan.lock
```
or
```shell
conan create . ... -b missing --lockfile-out mypackage.lock
conan upload -r my_remote --lockfile mypackage.lock
```
What do you think?

**Comment by memsharded** [comment]
Hi @Inujel 

Thanks for your suggestion. I think having some way to achieve that would make sense.
At the moment we don't plan to add this before 2.0, so I am putting it as a 2.X suggestion.
In the meantime, the best approach to have this would be:
- Create a Conan 2.0 custom ``upload`` command
- Use the PythonAPI to load the lockfile get the references from there, and call the upload API to upload it

I think this is doable in around 100 lines of python, the main issues right now would be the lack of documentation of the API, so reading the code and existing built-in commands would be the way.

**Comment by Inujel** [comment]
Thanks @memsharded !

> I think this is doable in around 100 lines of python

Yes indeed! Here's the script that I wrote for Conan 1.x in the meantime, if anyone's interested.

```py
from conans.client.conan_api import Conan
from conans.model.graph_lock import GraphLockFile
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('lockfile', nargs=1)
parser.add_argument('-r', '--remote', required=True)
args = parser.parse_args()

conan = Conan()
lockfile = GraphLockFile.load(args.lockfile[0], True)

for n in lockfile.graph_lock.nodes.values():
   if n.ref.revision and n.package_id:
      full_ref = f'{n.ref}#{n.ref.revision}'
      conan.upload(full_ref, package=n.package_id, remote_name=args.remote)
```

**Comment by memsharded** [comment]
Nice, many thanks!

Yes, in Conan 2.0, it might take a few more lines, because the api is more detailed, and it is not a single ``conan.upload()`` command, but it might require a few more calls.

**Comment by soerengustenhoff** [comment]
as described in https://github.com/conan-io/conan-package-tools/issues/616 we have the exact same issue.

We are subscribing to this issue to see where we are heading with conan 2.0 :)

**Comment by memsharded** [comment]
This feature has been finally implemented in https://github.com/conan-io/conan/pull/13928, will be in Conan 2.0.7

The flow from lockfile will be something like:
- ``conan install --lockfile=conan.lock --format=json > graph.json``
- ``conan list --graph=graph.json --format=json > pkglist.json``
- ``conan upload --list=pkglist.json -r=myremote``

Closing as implemented, looking forward your feedback after 2.0.7 (or now if you test it from release/2.0 branch)

**Comment by Inujel** [comment]
Amazing! Thank you so much!
